### ISSUE TITLE
0.9.8zd/1.0.0p Build fail in cryptlib.h/bio.h BIO_FLAGS_UPLINK macro redefinition

### ISSUE THREAD

**Issue by winkelsdorf** [openssl/openssl#210]
The Build fails in cryptlib.h/bio.h for 0.9.8zd and 1.0.0p with the following message:

1.0.0p
```
C:\OpenSSL\Temp\openssl-1.0.0p-x32\tmp32dll\cryptlib.h(68) : error C2220: warning treated as error - no 'object' file generated

C:\OpenSSL\Temp\openssl-1.0.0p-x32\tmp32dll\cryptlib.h(68) : warning C4005: 'BIO_FLAGS_UPLINK' : macro redefinition
C:\OpenSSL\Temp\openssl-1.0.0p-x32\inc32\openssl/bio.h(182) : see previous definition of 'BIO_FLAGS_UPLINK'
```

0.9.8zd
```
tmp32dll\cryptlib.h(68) : error C2220: warning treated as error - no 'object' file generated

tmp32dll\cryptlib.h(68) : warning C4005: 'BIO_FLAGS_UPLINK' : macro redefinition inc32\openssl/bio.h(180) : see previous definition of 'BIO_FLAGS_UPLINK'
```

Platform Win 7 x64. Building for Win32:
```
perl configure VC-WIN32
ms\do_nasm.bat
```

Previous versions built fine (0.9.8zc, 1.0.1o).

Using Microsoft Visual C++ 2008 Express Edition and The Netwide Assembler (NASM) v2.11.05 for Win32.

**Comment by NewEraCracker** [comment]
Also having this when building with Visual Studio 6 with:
```
perl Configure VC-WIN32
perl util\mkfiles.pl >MINFO
perl util\mk1mf.pl no-asm dll VC-WIN32 >32dll.mak
perl util\mkdef.pl 32 libeay > ms\libeay32.def
perl util\mkdef.pl 32 ssleay > ms\ssleay32.def
nmake -f 32dll.mak
```
Had to edit generated 32dll.mak and remove the `/WX` option from CFLAG as stated at INSTALL.W32

**Comment by NewEraCracker** [comment]
An even better fix.

Open `crypto/ecdsa/ecs_vrf.c`

Find
```
#include "ecs_locl.h"
#include "cryptlib.h"
```

Replace with
```
#include "cryptlib.h"
#include "ecs_locl.h"
```

It seems to be a bug in header include order.

**Comment by mattcaswell** [comment]
Either of the above solutions are acceptable workarounds for this problem.

**Comment by winkelsdorf** [comment]
Using the 2nd fix provided by NewEraCracker everything working as expected. Thank you!

I would suggest merging this into the two affected branches. I'll leave this issue, despite fixed, open until somebody merges the changes.

**Comment by mattcaswell** [comment]
There's an official fix going through review at the moment...hopefully should be there soon.

**Comment by winkelsdorf** [comment]
Ok, thanks for the follow-up. Nothing complicated here, just overlooked the include order in the aforementioned file, instead I was looking at the redeclaration itself.. Thanks for merging the changes!

**Comment by winkelsdorf** [comment]
Offtopic: By the way, as you are part of the official Team: I am providing pre-compiled Win32/64 libraries for about 10 years now, as I am part of a network component development team (Indy Components) for the Delphi Language.

The main difference between my releases and the shining light/slproweb releases (as linked at OpenSSL's binary distribution page https://www.openssl.org/related/binaries.html) is that my libraries are compiled without external dependencies to the Microsoft Visual Studio Runtime DLLs, expect for the system provided msvcrt.dll. Basically these precompiled versions allow an easier distribution without having to install corresponding, newer VC runtime dlls, too. 

Feel free to include a link to my blog and/or the Fulgan Mirror in Switzerland, where I am hosting them:
http://indy.fulgan.com/SSL/

**Comment by mattcaswell** [comment]
@winkelsdorf a link has been added on the binaries page

**Comment by mattcaswell** [comment]
The official fix for this issue is in commit 94e5cf36 for 1.0.0, and 56abaa14 for 0.9.8

**Comment by winkelsdorf** [comment]
Thanks for adding a link. I verified that both 9435cf3 and 56abaa1 are working correctly. 1.0.1l, 1.0.0q and 0.9.8ze are building fine without any modification. Thank you!
